client/lib/viewport/test/react-helpers.js

```diff
@@ -0,0 +1,268 @@
+/**
+ * @format
+ * @jest-environment jsdom
+ */
+
+/**
+ * Internal dependencies
+ */
+import React, { useState, useEffect } from 'react';
+import ReactDOM from 'react-dom';
+import { act } from 'react-dom/test-utils';
+
+let helpers;
+
+const listeners = {};
+
+const matchesMock = jest.fn( () => true );
+const addListenerMock = jest.fn( ( query, listener ) => {
+	if ( listeners[ query ] ) {
+		listeners[ query ].push( listener );
+	} else {
+		listeners[ query ] = [ listener ];
+	}
+} );
+const removeListenerMock = jest.fn( ( query, listener ) => {
+	if ( listeners[ query ] ) {
+		listeners[ query ] = listeners[ query ].filter( item => item !== listener );
+	}
+} );
+
+function callQueryListeners( query, value ) {
+	for ( const listener of listeners[ query ] ) {
+		listener( { matches: value } );
+	}
+}
+
+const matchMediaMock = jest.fn( query => {
+	const mediaListObjectMock = {
+		addListener: listener => addListenerMock( query, listener ),
+		removeListener: listener => removeListenerMock( query, listener ),
+	};
+	// Add matches read-only property.
+	Object.defineProperty( mediaListObjectMock, 'matches', {
+		get: () => matchesMock( query ),
+	} );
+	return mediaListObjectMock;
+} );
+
+describe( 'viewport/react-helpers', () => {
+	let container;
+
+	// Auxiliary method to test a valid component.
+	function runComponentTests( TestComponent, query ) {
+		// Test initial state (defaults to true).
+		act( () => {
+			ReactDOM.render(
+				<div>
+					<TestComponent />
+					<TestComponent />
+					<TestComponent />
+				</div>,
+				container
+			);
+		} );
+
+		expect( container.textContent ).toBe( 'truetruetrue' );
+		expect( listeners[ query ] ).not.toBe( undefined );
+		expect( listeners[ query ].length ).toBe( 3 );
+
+		// Simulate a window resize by calling the registered listeners for a query
+		// with a different value (false).
+		act( () => {
+			callQueryListeners( query, false );
+		} );
+
+		expect( container.textContent ).toBe( 'falsefalsefalse' );
+
+		// Ensure that listeners are cleaned up when the component unmounts.
+		act( () => {
+			ReactDOM.render( <div />, container );
+		} );
+
+		expect( listeners[ query ].length ).toBe( 0 );
+	}
+
+	// Auxiliary class for HOC tests.
+	class BaseComponent extends React.Component {
+		render() {
+			const isActive = this.props.isBreakpointActive;
+			return isActive ? 'true' : 'false';
+		}
+	}
+
+	beforeAll( async () => {
+		window.matchMedia = matchMediaMock;
+		helpers = await import( '../react-helpers' );
+		// Disable console warnings.
+		jest.spyOn( console, 'warn' ).mockImplementation( () => '' );
+	} );
+
+	beforeEach( () => {
+		container = document.createElement( 'div' );
```

Comment:
It would be nice to unmount the mounted components and remove the container element after each test.

When writing reactive tests with `i18n-calypso` in #31043, I was getting this warning:
```
Warning: An update to Label inside a test was not wrapped in act(...).

When testing, code that causes React state updates should be wrapped into act(...):

act(() => {
  /* fire events that update state */
});
/* assert on the output */
      
This ensures that you're testing the behavior the user would see in the browser. Learn more at https://fb.me/react-wrap-tests-with-act
    in Label
```

It was caused by two tests interacting:
```
beforeEach( () => {
  container = document.createElement( 'div' );
  document.body.appendChild( container );
} );

afterEach( () => {
  document.body.removeChild( container );
  container = null;
} );

test( () => {
  act( () => {
    ReactDOM.render( <Label />, container );
  } );
} );

test( () => {
  // reset locale to default
  i18n.setLocale();

  act( () => {
    ReactDOM.render( <Label />, container );
  } );
}
```

It turned out that when the second test called `i18n.setLocale()`, the component from the first test was still mounted and reacted to updates!

Adding `ReactDOM.unmountComponentAtNode( container )` to `afterEach` fixed that.

Reply:
Great idea! It's always better to run tests on a node that's actually attached to the DOM, but that does require the cleanup you mentioned. That's all in, now.

Reply:
The React docs at https://reactjs.org/docs/test-utils.html#act say that the DOM node needs to attached to a document in order to receive any DOM events. We are lucky that the tests don't dispatch any.